<a href="https://colab.research.google.com/github/dwyuliaan/dataset_modul_4/blob/main/%5BRumpang%5D_Final_Project.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Business Understanding

Anggaplah kita merupakan seorang Data Scientist dari perusahaan e-commerce Tokopedia, perusahaan sudah menguji coba 5 jenis kampanye pemasaran ( promosi ) yang berbeda di satu waktu secara bersamaan di bulan lalu. Saat ini, perusahaan ingin melihat promosi mana yang paling sering diikuti oleh customer.

Perusahaan juga membutuhkan mesin prediksi apakah customer yang datang dan beraktivitas di website perusahaan berpeluang menggunakan jenis promosi yang mana antara promosi ke 1 - 5. Agar, perusahaan tidak membuang - buang biaya promosi jika kelima jenis promosi tersebut terus di promosikan setiap hari, dan tentu saja untuk menghindari kerugian besar namun tetap mempertahankan customer yang setia serta menarik customer baru.

Anggap saja kolom campaign ke- :     

1. 1 = diskon 50%
2. 2 = beli 1 gratis 1
3. 3 = flash sale
4. 4 = free ongkir
5. 5 = cashback 30%

# Data Understanding

- **registration_id:** *ID unik yang digunakan untuk mengidentifikasi pelanggan.*

Format: String.

Contoh: RK20202344.

- **name:** *Nama lengkap pelanggan.*

Format: String.

Contoh: Fakhrul Fandy Widyaningtias.

- **email:** *Alamat email pelanggan.*

Format: String.

Contoh: fakhrulfandywidyaningtias@outlook.com.

- **phone:** *Nomor telepon pelanggan.*

Format: Float (dengan angka besar, mungkin perlu format ulang untuk menyimpan
tanpa notasi ilmiah).

Contoh: 6285244221090.

- **age:** *Usia pelanggan.*

Format: Float.

Contoh: 27.0.

- **income_group:** *Kelompok penghasilan pelanggan (kategori numerik).*

Format: Integer.

Contoh: 2 (kelompok penghasilan kedua).

- **active:** *Status aktif pelanggan (apakah pelanggan aktif).*

Format: Boolean.

Contoh: True.

- **city:** *Kota tempat tinggal pelanggan.*

Format: String.

Contoh: Tanjungpinang.

- **province:** *Provinsi tempat tinggal pelanggan.*

Format: String.

Contoh: Kepulauan Riau.

- **hobby:** *Hobi pelanggan.*

Format: String.

Contoh: Amateur radio.

- **fav_movie:** *Film favorit pelanggan.*

Format: String.

Contoh: The Godfather.

- **website_visit_count:** *Jumlah kunjungan pelanggan ke website.*

Format: Integer.

Contoh: 51.

- **website_recency:** *Jumlah hari sejak terakhir kali pelanggan mengunjungi website.*

Format: Integer.

Contoh: 75.

- **website_purchase_rate:** *Rasio pembelian pelanggan terhadap kunjungan website.*

Format: Float.

Contoh: 0.68 (68%).

- **website_purchase_amount:** *Total nilai pembelian pelanggan di website.*

Format: Integer.

Contoh: 612000 (Rp 612.000).

- **website_cs_call_count:** *Jumlah panggilan pelanggan ke customer service.*

Format: Integer.

Contoh: 1.

- **accept_campaign_1 hingga accept_campaign_5:** *Status apakah pelanggan menerima atau berpartisipasi dalam kampanye pemasaran tertentu.*

Format: Boolean.

Contoh: True (menerima kampanye).

# Data Preparation

In [ ]:
import pandas as pd
df = pd.read_csv("https://raw.githubusercontent.com/krishna0604/dataset_advanced/refs/heads/main/data_customers_csv.csv")

In [ ]:
df.head()

registration_id                         name  \
0      RK20202344  Fakhrul Fandy Widyaningtias   
1      RK20202345          Ferdiansyah Amalina   
2      RK20202346            Ficky Rizkyananta   
3      RK20202347             Singgih Kharisma   
4      RK20202348             Jeremiah Dinanti   

                                   email         phone   age  income_group  \
0  fakhrulfandywidyaningtias@outlook.com  6.285244e+09  27.0             2   
1           ferdiansyahamalina@yahoo.com  6.285245e+09  31.0             3   
2           fickyrizkyananta@hotmail.com  6.285245e+09  15.0             1   
3            singgihkharisma@hotmail.com  6.285245e+09  49.0             4   
4                    jeremia@rakamin.com  6.285245e+09  25.0             2   

   active                             city        province          hobby  \
0   False                    Tanjungpinang  Kepulauan Riau  Amateur radio   
1    True                            Bogor      Jawa Barat      Animation   
2    True  Kota Administrasi Jakarta Pusat         Jakarta          Anime   
3    True                            Metro         Lampung    Aquascaping   
4   False                           Malang      Jawa Timur            Art   

   ... website_visit_count  website_recency  website_purchase_rate  \
0  ...                  51               75                   0.68   
1  ...                  88                3                   0.51   
2  ...                  30               19                   0.69   
3  ...                  63                7                   0.80   
4  ...                  99               81                   0.15   

   website_purchase_amount  website_cs_call_count  accept_campaign_1  \
0                   612000                      1               True   
1                    51000                      3               True   
2                   276000                      1              False   
3                     3200                      3              False   
4                   105000                      2               True   

   accept_campaign_2  accept_campaign_3  accept_campaign_4  accept_campaign_5  
0              False               True               True              False  
1              False              False               True              False  
2               True              False               True              False  
3               True               True               True               True  
4               True               True               True               True  

[5 rows x 21 columns]

In [ ]:
df.shape

(107, 21)

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 107 entries, 0 to 106
Data columns (total 21 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   registration_id          107 non-null    object 
 1   name                     107 non-null    object 
 2   email                    107 non-null    object 
 3   phone                    104 non-null    float64
 4   age                      106 non-null    float64
 5   income_group             107 non-null    int64  
 6   active                   107 non-null    bool   
 7   city                     107 non-null    object 
 8   province                 107 non-null    object 
 9   hobby                    102 non-null    object 
 10  fav_movie                69 non-null     object 
 11  website_visit_count      107 non-null    int64  
 12  website_recency          107 non-null    int64  
 13  website_purchase_rate    107 non-null    float64
 14  website_purchase_amount  1

Insight :    

1. Terdapat 107 baris dan 21 Kolom

2. Terdapat Missing Data pada kolom phone (3 values), age (1 values), hobby (5 values), dan fav_movie (38 values)

In [ ]:
df.describe()

phone         age  income_group  website_visit_count  \
count  1.040000e+02  106.000000    107.000000           107.000000   
mean   6.123076e+09   33.132075      2.803738            51.831776   
std    9.515607e+08   11.480895      1.216547            29.269562   
min    6.284456e+08   15.000000      1.000000             5.000000   
25%    6.285234e+09   24.250000      2.000000            24.500000   
50%    6.286627e+09   33.000000      3.000000            49.000000   
75%    6.287881e+09   43.750000      4.000000            74.000000   
max    6.289633e+09   57.000000      5.000000           100.000000   

       website_recency  website_purchase_rate  website_purchase_amount  \
count       107.000000             107.000000             1.070000e+02   
mean         20.130841               0.675794             8.322712e+05   
std          18.765225               0.243338             1.713776e+06   
min           0.000000               0.000000             0.000000e+00   
25%           7.500000               0.610000             1.125000e+05   
50%          17.000000               0.720000             3.400000e+05   
75%          25.000000               0.830000             6.265000e+05   
max          93.000000               1.000000             9.500000e+06   

       website_cs_call_count  
count             107.000000  
mean                0.663551  
std                 1.132237  
min                 0.000000  
25%                 0.000000  
50%                 0.000000  
75%                 1.000000  
max                 3.000000

In [ ]:
for i in df:
  print(f'''nilai unik dari kolom {i}''')
  print(df[i].unique())
  print()

nilai unik dari kolom registration_id
['RK20202344' 'RK20202345' 'RK20202346' 'RK20202347' 'RK20202348'
 'RK20202349' 'RK20202350' 'RK20202351' 'RK20202352' 'RK20202353'
 'RK20202354' 'RK20202355' 'RK20202356' 'RK20202357' 'RK20202358'
 'RK20202359' 'RK20202360' 'RK20202361' 'RK20202362' 'RK20202363'
 'RK20202364' 'RK20202365' 'RK20202366' 'RK20202367' 'RK20202368'
 'RK20202369' 'RK20202370' 'RK20202371' 'RK20202372' 'RK20202373'
 'RK20202374' 'RK20202375' 'RK20202376' 'RK20202377' 'RK20202378'
 'RK20202379' 'RK20202380' 'RK20202381' 'RK20202382' 'RK20202383'
 'RK20202384' 'RK20202385' 'RK20202386' 'RK20202387' 'RK20202388'
 'RK20202389' 'RK20202390' 'RK20202391' 'RK20202392' 'RK20202393'
 'RK20202394' 'RK20202395' 'RK20202396' 'RK20202397' 'RK20202398'
 'RK20202399' 'RK20202400' 'RK20202401' 'RK20202402' 'RK20202403'
 'RK20202404' 'RK20202405' 'RK20202406' 'RK20202407' 'RK20202408'
 'RK20202409' 'RK20202410' 'RK20202411' 'RK20202412' 'RK20202413'
 'RK20202414' 'RK20202415' 'RK20202416

Insight :    

1. Customer berada pada rentang usia 15 - 57 tahun.

2. Income Group terdiri dari 5 grup, 1 - 5

3. Kunjungan ke website pada rentang 0 - 100 kali dengan rata" nya adalah 51 kali ( make sense )

4. Jumlah hari sejak terakhir kali berkunjung ke website ( website_recency ) adalah 0 - 93 hari.

## Data Cleaning

### a.) Handle Missing Values

Sebelumnya telah diketahui bahwa : Terdapat Missing Data pada kolom phone (3 values), age (1 values), dan hobby (5 values), dan fav_movie (38 values)

In [ ]:
df.isnull().sum()

registration_id             0
name                        0
email                       0
phone                       3
age                         1
income_group                0
active                      0
city                        0
province                    0
hobby                       5
fav_movie                  38
website_visit_count         0
website_recency             0
website_purchase_rate       0
website_purchase_amount     0
website_cs_call_count       0
accept_campaign_1           0
accept_campaign_2           0
accept_campaign_3           0
accept_campaign_4           0
accept_campaign_5           0
dtype: int64

Handle dengan :    

1. Menghapus baris phone, karena kemungkinan customer tanpa menggunakan nomor telepon adalah customer yang tidak terlalu dipercaya + menhapus 3 dari 107 baris tidak akan menghilangkan banyak infromasi

2. Mengisi null value pada kolom age dengan rata" karena missing valuenya hanya ada 1 dari 107 baris

3. Mengisi null value pada kolom hobby dengan modus karena tipe datanya adalah object

4. Menghapus kolom fav_movie karena nama film tidak ada hubungannya dengan daya beli customer dan akan sangat berpengaruh jika null values tersebut dihapus yang berarti harus menghapus 38 dari 107 baris dan akan mengurangi informasi.

Handle Missing Values Phone :

Insight : Karena kebetulan baris kolom age yang kosong juga merupakan informasi dari baris kolom phone yang kosong dan ikut terhapus, maka handle missing value kolom age tidak lagi diperlukan.

Handle Missing Values kolom Hobby :

Insight : Ghost Hunting adalah hobby yang paling banyak disukai, maka Ghost Hunting adalah modus.

Menghapus kolom fav_movie

In [ ]:
print("----------------")
df.shape

----------------


(104, 20)

Hasil akhir dari handle missing values adalah berkurangnya baris dari 107 baris menjadi 104. Hal ini tidak merubah banyak infromasi pada data dan data sudah bersih dari missng values

### b.) Handle Duplicates

In [ ]:
df.duplicated().sum()

6

In [ ]:
for i in df:
  print(f'''duplikasi dari kolom {i}''')
  print(df.loc[df[i].duplicated()])
  print()

duplikasi dari kolom registration_id
   registration_id                     name  \
8       RK20202351  Dikposa Yosua Krisnanto   
39      RK20202381              Jova Imania   
42      RK20202383          Garin Sugiyanto   
52      RK20202392             Sigit Amalia   
92      RK20202431           Prima Hermawan   
93      RK20202431           Prima Hermawan   

                                email         phone   age  income_group  \
8   dikposayosuakrisnanto@hotmail.com  6.281907e+09  28.0             2   
39               jovaimania@gmail.com  6.284688e+08  17.0             1   
42           garinsugiyanto@yahoo.com  6.287881e+09  33.0             3   
52          sigitamalia@roketmail.com  6.287893e+09  48.0             4   
92       primahermawan@rocketmail.com  6.281856e+09  50.0             5   
93       primahermawan@rocketmail.com  6.281856e+09  50.0             5   

    active                             city          province           hobby  \
8    False  Kota Administr

Betul, terdapat 6 kolom yang memiliki values dan informasi yang sama sehingga dianggap sebagai kolom yang berduplikat. Kita ambil contoh kolom nama dengan nama Dikposa Yosua Krisnanto

In [ ]:
df.loc[df["name"]=="Dikposa Yosua Krisnanto"]

registration_id                     name                              email  \
7      RK20202351  Dikposa Yosua Krisnanto  dikposayosuakrisnanto@hotmail.com   
8      RK20202351  Dikposa Yosua Krisnanto  dikposayosuakrisnanto@hotmail.com   

          phone   age  income_group  active                             city  \
7  6.281907e+09  28.0             2   False  Kota Administrasi Jakarta Pusat   
8  6.281907e+09  28.0             2   False  Kota Administrasi Jakarta Pusat   

  province           hobby  website_visit_count  website_recency  \
7  Jakarta  Fingerpainting                   70               57   
8  Jakarta  Fingerpainting                   70               57   

   website_purchase_rate  website_purchase_amount  website_cs_call_count  \
7                    0.0                        0                      0   
8                    0.0                        0                      0   

   accept_campaign_1  accept_campaign_2  accept_campaign_3  accept_campaign_4  \
7              False              False              False              False   
8              False              False              False              False   

   accept_campaign_5  
7              False  
8              False

Karena ada 2 baris dengan value yang sama, maka akan dihapus salah satunya ( indeks 8 dihapus,ke-7 tidak ). Oleh karena itu, handle duplicates sangat penting dan tidak akan mengubah informasi apapun ketika baris duplikat dihapus

In [ ]:
df.loc[df["name"]=="Dikposa Yosua Krisnanto"]

registration_id                     name                              email  \
7      RK20202351  Dikposa Yosua Krisnanto  dikposayosuakrisnanto@hotmail.com   

          phone   age  income_group  active                             city  \
7  6.281907e+09  28.0             2   False  Kota Administrasi Jakarta Pusat   

  province           hobby  website_visit_count  website_recency  \
7  Jakarta  Fingerpainting                   70               57   

   website_purchase_rate  website_purchase_amount  website_cs_call_count  \
7                    0.0                        0                      0   

   accept_campaign_1  accept_campaign_2  accept_campaign_3  accept_campaign_4  \
7              False              False              False              False   

   accept_campaign_5  
7              False

In [ ]:
df.duplicated().sum()

0

## Data Visualization

Setelah data sudah bersih, maka saatnya untuk mengambil insight lebih banyak menggunakan visualisasi data. Karena di dalam data ini tidak ada kolom / variabel tanggal-bulan-tahun (time series) maka line chart tidak akan diaplikasikan

Sesuai dengan informasi pada bagian Business Understanding : perusahaan sudah menguji coba 5 jenis kampanye pemasaran ( promosi ) yang berbeda di satu waktu secara bersamaan di bulan lalu. Saat ini, perusahaan ingin melihat promosi mana yang paling sering diikuti oleh customer. Maka saatnya untuk membuat visualisasinya

Insight = Campaign yang paling disukai oleh customer adalah campaign nomor 3 ( flash sale ) dan campaign 5 ( cashback 30% ) adalah campaign yang paling tidak disukai oleh customer

### BONUS

Buatlah visualisasi 3 provinsi mana yang memiliki customer dengan status aktif berbelanja di perusahaan

3 Provinsi dengan jumlah customer aktif paling banyak adalah : Jawa Barat, Jawa Timur, dan Jawa Tengah

## Data Manipulation

### a.) Binning

Binning digunakan untuk mengubah data numerik menjadi kategorik dengan tujuan untuk membuat data menjadi lebih simpel.

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 98 entries, 0 to 106
Data columns (total 20 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   registration_id          98 non-null     object 
 1   name                     98 non-null     object 
 2   email                    98 non-null     object 
 3   phone                    98 non-null     float64
 4   age                      98 non-null     float64
 5   income_group             98 non-null     int64  
 6   active                   98 non-null     bool   
 7   city                     98 non-null     object 
 8   province                 98 non-null     object 
 9   hobby                    98 non-null     object 
 10  website_visit_count      98 non-null     int64  
 11  website_recency          98 non-null     int64  
 12  website_purchase_rate    98 non-null     float64
 13  website_purchase_amount  98 non-null     int64  
 14  website_cs_call_count    98 non-

Mengelompokkan kolom age, dengan kelompok produktif, dewasa, dan paruh baya ( 40 - 59 ) namun harus mengetahui rentang umur terlebih dahulu. Rentang umur bisa dicari menggunakan binning

In [ ]:
pd.qcut(df["age"], 3).value_counts()

age
(14.999, 27.0]    36
(38.0, 57.0]      32
(27.0, 38.0]      30
Name: count, dtype: int64

### b.) Clipping ( Handle Outlier )

In [ ]:
df.describe()

phone        age  income_group  website_visit_count  \
count  9.800000e+01  98.000000     98.000000            98.000000   
mean   6.170918e+09  32.693878      2.734694            51.336735   
std    8.041015e+08  11.348653      1.180171            29.288450   
min    6.284456e+08  15.000000      1.000000             5.000000   
25%    6.285245e+09  24.000000      2.000000            23.250000   
50%    6.287304e+09  32.000000      3.000000            49.500000   
75%    6.287881e+09  42.000000      4.000000            73.750000   
max    6.289633e+09  57.000000      5.000000           100.000000   

       website_recency  website_purchase_rate  website_purchase_amount  \
count        98.000000              98.000000             9.800000e+01   
mean         20.316327               0.686531             8.663471e+05   
std          19.051529               0.233521             1.783862e+06   
min           0.000000               0.000000             0.000000e+00   
25%           7.250000               0.610000             1.087500e+05   
50%          17.000000               0.720000             3.285000e+05   
75%          25.000000               0.845000             6.282500e+05   
max          93.000000               1.000000             9.500000e+06   

       website_cs_call_count  
count              98.000000  
mean                0.693878  
std                 1.143409  
min                 0.000000  
25%                 0.000000  
50%                 0.000000  
75%                 1.000000  
max                 3.000000

insight ( Praduga ) :    

1. Kolom website_recency kemungkinan terdapat high outlier karena rentangnya adalah dari 0 - 93 namun rata-ratanya adalah 20.

2. Kolom website_purchase_amount kemungkinan terdapat outlier karena rentangnya adalah dari 0 - 9.5jt namun rata-ratanya adalah 866rb dengan median 328rb.

Info tambahan :    

e adalah eksponen atau dikali dengan 10 pangkat. jadi e+05 artinya dikali 10 pangkat 5. yang berarti, rentang website_purchase_amount adalah dari 0 - 9.500.000 dan mean nya adalah 866.000

Handle outlier website_recency :

Kolom website_recency sudah tidak ada outlier

Handle outlier website_purchase_amount

kolom website_purchase_amount yang baru adalah rentang dari 0 - 1.4jt dengan rata-ratanya 437rb ( make sense )

### c.) Scaling

In [ ]:
df.describe()

phone        age  income_group  website_visit_count  \
count  9.800000e+01  98.000000     98.000000            98.000000   
mean   6.170918e+09  32.693878      2.734694            51.336735   
std    8.041015e+08  11.348653      1.180171            29.288450   
min    6.284456e+08  15.000000      1.000000             5.000000   
25%    6.285245e+09  24.000000      2.000000            23.250000   
50%    6.287304e+09  32.000000      3.000000            49.500000   
75%    6.287881e+09  42.000000      4.000000            73.750000   
max    6.289633e+09  57.000000      5.000000           100.000000   

       website_recency  website_purchase_rate  website_purchase_amount  \
count        98.000000              98.000000             9.800000e+01   
mean         20.316327               0.686531             8.663471e+05   
std          19.051529               0.233521             1.783862e+06   
min           0.000000               0.000000             0.000000e+00   
25%           7.250000               0.610000             1.087500e+05   
50%          17.000000               0.720000             3.285000e+05   
75%          25.000000               0.845000             6.282500e+05   
max          93.000000               1.000000             9.500000e+06   

       website_cs_call_count  Clipped_website_recency  \
count              98.000000                98.000000   
mean                0.693878                18.367347   
std                 1.143409                13.423668   
min                 0.000000                 0.000000   
25%                 0.000000                 7.250000   
50%                 0.000000                17.000000   
75%                 1.000000                25.000000   
max                 3.000000                51.625000   

       Clipped_website_purchase_amount  
count                     9.800000e+01  
mean                      4.378522e+05  
std                       4.109651e+05  
min                       0.000000e+00  
25%                       1.087500e+05  
50%                       3.285000e+05  
75%                       6.282500e+05  
max                       1.407500e+06

kita mengabaikan kolom "phone" karena semua baris memiliki nomor telepon yang berbeda-beda kecuali jika ada duplikasi, dan kita sudah handle diatas. sekarang kita analisis apakah skala nya merata dari kolom selain kolom phone :     

1. age = 15 - 57
2. income_group = 1 - 5 ( termasuk kategori, jadi tidak usah di scaling )
3. website_visit_count = 5 - 100
4. website_recency = 0 - 93
5. website_purchase_rate = 0 - 1 (persen)
6. website_purchase_amount = 0 - 9.5jt
7. website_cs_call_count = 0 - 3
8. Clipped_website_recency = 0 - 51
9. Clipped_website_purchase_amount = 0 - 1.4jt

Berdasarkan informasi diatas, sudah jelas yang akan kita scaling adalah kolom website_purchase_amount dan Clipped_website_purchase_amount. Namun kita akan membiarkan kolom website_purchase_amount karena sudah digantikan dengan yang tanpa outlier pada kolom Clipped_website_purchase_amount sehingga tidak akan digunakan untuk modelling.

Kita akan melakukan scaling dengan metode StandardScaler karena kolom ini merupakan kolom harga dan tidak menggunakan MinMaxScaler karena tidak ingin mengubah value dari 0 - 1.

std_clipped_website_purchase_amount

In [ ]:
df["std_clipped_website_purchase_amount"].describe()

count    9.800000e+01
mean     4.304946e-17
std      1.005141e+00
min     -1.070902e+00
25%     -8.049206e-01
50%     -2.674545e-01
75%      4.656762e-01
max      2.371571e+00
Name: std_clipped_website_purchase_amount, dtype: float64

### d.) Encoding

In [ ]:
for i in df:
  print(f'''nilai unik dari kolom {i}''')
  print(df[i].unique())
  print()

nilai unik dari kolom registration_id
['RK20202344' 'RK20202345' 'RK20202346' 'RK20202347' 'RK20202348'
 'RK20202349' 'RK20202350' 'RK20202351' 'RK20202352' 'RK20202353'
 'RK20202354' 'RK20202355' 'RK20202356' 'RK20202358' 'RK20202359'
 'RK20202360' 'RK20202361' 'RK20202362' 'RK20202363' 'RK20202364'
 'RK20202365' 'RK20202366' 'RK20202368' 'RK20202369' 'RK20202370'
 'RK20202371' 'RK20202372' 'RK20202373' 'RK20202375' 'RK20202376'
 'RK20202377' 'RK20202378' 'RK20202379' 'RK20202380' 'RK20202381'
 'RK20202382' 'RK20202383' 'RK20202384' 'RK20202385' 'RK20202386'
 'RK20202387' 'RK20202388' 'RK20202389' 'RK20202390' 'RK20202391'
 'RK20202392' 'RK20202393' 'RK20202394' 'RK20202395' 'RK20202396'
 'RK20202397' 'RK20202398' 'RK20202399' 'RK20202400' 'RK20202401'
 'RK20202402' 'RK20202403' 'RK20202404' 'RK20202405' 'RK20202406'
 'RK20202407' 'RK20202408' 'RK20202409' 'RK20202410' 'RK20202411'
 'RK20202412' 'RK20202413' 'RK20202414' 'RK20202415' 'RK20202416'
 'RK20202417' 'RK20202418' 'RK20202419

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 98 entries, 0 to 106
Data columns (total 24 columns):
 #   Column                               Non-Null Count  Dtype  
---  ------                               --------------  -----  
 0   registration_id                      98 non-null     object 
 1   name                                 98 non-null     object 
 2   email                                98 non-null     object 
 3   phone                                98 non-null     float64
 4   age                                  98 non-null     float64
 5   income_group                         98 non-null     int64  
 6   active                               98 non-null     bool   
 7   city                                 98 non-null     object 
 8   province                             98 non-null     object 
 9   hobby                                98 non-null     object 
 10  website_visit_count                  98 non-null     int64  
 11  website_recency                      9

Kolom registration_id, name, email, dan phone tidak mungkin di encoding dan tidak akan dijadikan sebagai variabel prediktor pada model karena valuenya unik disetiap baris sehingga tidak akan ditemukan korelasi dan tidak bermanfaat.

Kolom hobby juga tidak akan di encoding dan dijadikan variabel prediktor karena terlalu banyak unique values sehingga tidak berkorelasi. Sama halnya dengan kolom city

**Kolom yang akan di encoding adalah kolom province dan age_group, namun akan kita kelompokkan terlebih dahulu kolom province menjadi kelompok indo_timur, indo_barat, dan indo_tengah.**

Kolom yang lain tidak di encoding karena sudah dalam bentuk angka dan bisa dibaca oleh mesin



Encoding kolom province_group dan age_group

Sebelum melakukan modelling, pastikan kita sudah mengambil kolom" yang SIAP untuk dijadikan sebagai variabel prediktor dan target. AYO KITA FILTER !

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 98 entries, 0 to 106
Data columns (total 25 columns):
 #   Column                               Non-Null Count  Dtype  
---  ------                               --------------  -----  
 0   registration_id                      98 non-null     object 
 1   name                                 98 non-null     object 
 2   email                                98 non-null     object 
 3   phone                                98 non-null     float64
 4   age                                  98 non-null     float64
 5   income_group                         98 non-null     int64  
 6   active                               98 non-null     bool   
 7   city                                 98 non-null     object 
 8   province                             98 non-null     object 
 9   hobby                                98 non-null     object 
 10  website_visit_count                  98 non-null     int64  
 11  website_recency                      9

Kita tidak akan mengambil kolom :    

1. registration_id, name, email, phone, age ( sudah ada age_group yang di encoding ), city, province ( sudah ada province_group yang di encoding ), hobby, website_recency ( ssudah ada Clipped_website_recency ), website_purchase_amount serta Clipped_website_purchase_amount ( sudah ada std_clipped_website_purchase_amount ), age_group dan province_group(sudah di encoding)

# Modelling

## Modelling preparation

Kita mempunyai 1 Task yang diberikan, yaitu membuat mesin prediksi apakah customer yang datang dan beraktivitas di website perusahaan berpeluang menggunakan jenis promosi yang mana antara promosi ke 1 - 5.

Kita menggunakan hasil_df, bukan df

membagi data menjadi data train dan data test

## Modelling

In [ ]:
hasil_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 98 entries, 0 to 106
Data columns (total 18 columns):
 #   Column                               Non-Null Count  Dtype  
---  ------                               --------------  -----  
 0   income_group                         98 non-null     int64  
 1   active                               98 non-null     bool   
 2   website_visit_count                  98 non-null     int64  
 3   website_purchase_rate                98 non-null     float64
 4   website_cs_call_count                98 non-null     int64  
 5   accept_campaign_1                    98 non-null     bool   
 6   accept_campaign_2                    98 non-null     bool   
 7   accept_campaign_3                    98 non-null     bool   
 8   accept_campaign_4                    98 non-null     bool   
 9   accept_campaign_5                    98 non-null     bool   
 10  Clipped_website_recency              98 non-null     float64
 11  std_clipped_website_purchase_amount  9

In [ ]:
X_test[2:3]

income_group  active  website_visit_count  website_purchase_rate  \
49             2    True                   69                   0.83   

    website_cs_call_count  Clipped_website_recency  \
49                      0                     22.0   

    std_clipped_website_purchase_amount  province_group_indo_barat  \
49                             2.371571                        1.0   

    province_group_indo_tengah  province_group_indo_timur  age_group_dewasa  \
49                         0.0                        0.0               0.0   

    age_group_paruh_baya  age_group_produktif  
49                   1.0                  0.0

Belum terdapat kolom - kolom target, yaitu "accept_campaign_1", "accept_campaign_2", "accept_campaign_3", "accept_campaign_4", "accept_campaign_5" karena merupakan data variabel prediktor test, khusunya pada data test indeks ke-2

In [ ]:
dt.predict(X_test[2:3])

array([[False, False,  True,  True,  True]])

Artinya mesin memprediksi bahwa pelanggan dengan indeks ke-2 pada data test adalah pelanggan yang diprediksi suka menggunakan rpomo / campaign ke 3,4,5 atau promo flash sale, free ongkir, dan cashback 30%

## Modelling Evaluation

In [ ]:
# import libraries
from sklearn.metrics import mean_absolute_error, accuracy_score, precision_score, recall_score, silhouette_score

recall dan precision menggunakan parameter micro. micro: Menghitung metrik global dengan meringkas semua label.

Jika hasil akurasi, precision, recall dirasa kurang memuaskan maka dianjurkan untuk melakukan data addition dengan manipulasi data, seperti melakukan :    

1. Encoding kolom income_group dan active
2. Melakukan pengelompokkan ( Binning )pada kolom website_visit_count ( misal megelompokkan menjadi jarang, sering, dan sangat sering ) lalu di encoding lagi hasilnya agar bisa dibaca oleh mesin
3. Melakukan pengelompokkan pada kolom city ( misal megelompokkan menjadi kota_barat, kota_tengah, dan kota timur ) lalu di encoding lagi hasilnya agar bisa dibaca oleh mesin
4. dst.